server/segment_rpc.go

```diff
@@ -477,31 +485,31 @@ func newBalanceUpdate(sess *BroadcastSession) (*BalanceUpdate, error) {
 	return update, nil
 }
 
-// func completeBalanceUpdate(sess *BroadcastSession, update *BalanceUpdate) {
-// 	if sess.Balance == nil {
-// 		return
-// 	}
-
-// 	// If the update's credit has not been spent then add the existing credit
-// 	// back to the balance
-// 	if update.Status == Staged {
-// 		sess.Balance.Credit(update.ExistingCredit)
-// 		return
-// 	}
-
-// 	// If the update did not include a processed debit then no change was received
-// 	// so we exit without updating the balance because the credit was spent already
-// 	if update.Status != ReceivedChange {
-// 		return
-// 	}
-
-// 	credit := new(big.Rat).Add(update.ExistingCredit, update.NewCredit)
-// 	// The change could be negative if the debit > credit
-// 	change := credit.Sub(credit, update.Debit)
-
-// 	// If the change is negative then this is equivalent to debiting abs(change)
-// 	sess.Balance.Credit(change)
-// }
+func completeBalanceUpdate(sess *BroadcastSession, update *BalanceUpdate) {
+	if sess.Balance == nil {
+		return
+	}
+
+	// If the update's credit has not been spent then add the existing credit
+	// back to the balance
+	if update.Status == Staged {
+		sess.Balance.Credit(update.ExistingCredit)
+		return
+	}
+
+	// If the update did not include a processed debit then no change was received
+	// so we exit without updating the balance because the credit was spent already
+	if update.Status != ReceivedChange {
```

Comment:
Doesn't this essentially zero the entire balance ? Why is there no residual balance possible at this point? Thanks.

Reply:
Yeah this would zero the entire balance. The motivation behind this is that if the segment is submitted (i.e. the `update.Status = CreditSpent`), but B does not receive a successful response before `SubmitSegment()` errors out then B does not know with certainty whether O successfully received the payment and applied credit/debit to its view of B's balance. If we instead add `existingCredit` back to B's balance then it is possible that O did receive the payment and applied credit/debit such that its view of B's balance is `< existingCredit`. Then, if B reuses the session with O, B would receive an insufficient balance error from O because B generates a payment based on an out of sync view of its balance. Considering the existing credit as "spent" in this scenario makes the workflow between B & O less dependent on the two parties having the same exact view of B's balance.

I actually just realized that technically we already zero the balance when `SubmitSegment()` encounters an error because we call [removeSession()](https://github.com/livepeer/go-livepeer/blob/master/server/broadcast.go#L92) which will clear the balance for the session. However, `removeSession()` will actually delete the manifestID key in the underlying `core.Balances` type. Meanwhile, if `update.Status = CreditSpent` when we complete the balance update in `SubmitSegment()`, the balance stored for the manifestID in the underlying `core.Balances` type will be left as `big.NewRat(0, 1)` (and then the entire map entry would be deleted when we call `removeSession()`). I'm ok with leaving this behavior as-is since `removeSession()` actually deletes the keys from `core.Balances`